pallets/pallet-dip-receiver/src/origin.rs

```diff
@@ -0,0 +1,89 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2023 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{traits::EnsureOrigin, RuntimeDebug};
+use scale_info::TypeInfo;
+use sp_std::marker::PhantomData;
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct DipOrigin<Identifier, AccountId> {
+	pub identifier: Identifier,
+	pub account_address: AccountId,
+}
+
+pub struct EnsureDipOrigin<Identifier, AccountId>(PhantomData<(Identifier, AccountId)>);
+
+#[cfg(not(feature = "runtime-benchmarks"))]
+impl<OuterOrigin, Identifier, AccountId> EnsureOrigin<OuterOrigin> for EnsureDipOrigin<Identifier, AccountId>
+where
+	OuterOrigin: From<DipOrigin<Identifier, AccountId>> + Into<Result<DipOrigin<Identifier, AccountId>, OuterOrigin>>,
+{
+	type Success = DipOrigin<Identifier, AccountId>;
+
+	fn try_origin(o: OuterOrigin) -> Result<Self::Success, OuterOrigin> {
+		o.into()
+	}
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+fn try_successful_origin<OuterOrigin, Identifier, AccountId>() -> Result<OuterOrigin, ()>
+where
+	OuterOrigin: From<DipOrigin<Identifier, AccountId>> + Into<Result<DipOrigin<Identifier, AccountId>, OuterOrigin>>,
+	Identifier: From<[u8; 32]>,
+	AccountId: From<[u8; 32]>,
+{
+	Ok(OuterOrigin::from(DipOrigin {
+		identifier: Identifier::from([0u8; 32]),
+		account_address: AccountId::from([0u8; 32]),
+	}))
+}
```

Comment:
why not move this into the trait?

Reply:
but you have the same trait bounds on the trait? I don't see the difference.

Reply:
Yes, you are right! 😁 I used a similar but apparently for a different reason in the past. In this case it does not make sense. Fixed in https://github.com/KILTprotocol/kilt-node/pull/489/commits/4d8f14d46af5fc40a164471809dee81cc87c6a99.